File: sim/env_reward.lua
-- Named learning-environment reward channels.
--
-- There is no implicit scalar reward. Every channel is registered with an
-- explicit role: `objective` channels are competition targets, `shaping`
-- channels must be preregistered and are always reported separately so an
-- ablation is a subtraction rather than a rerun, and `diagnostic` channels are
-- evaluation data that may never be optimized. The #128 fun-proxy metric family
-- is a diagnostic channel: selecting it as an objective or shaping term is a
-- validation error, and no channel may be named `fun`.

---@alias EnvRewardRole "objective"|"shaping"|"diagnostic"

---@alias EnvRewardChannelId
---| "match_outcome"
---| "goal_scored"
---| "goal_conceded"
---| "goal_difference_delta"
---| "possession_gain"
---| "shot_attempt"
---| "equipment_contact"
---| "experience_proxy_metrics"

---@alias EnvRewardErrorCode "malformed"|"unknown_channel"|"wrong_role"|"duplicate_channel"

---@class EnvRewardChannel
---@field id EnvRewardChannelId
---@field role EnvRewardRole
---@field optimizable boolean -- False for diagnostics: never a training target.
---@field preregistered boolean -- Shaping terms are declared here, not invented per run.
---@field description string

---@class EnvRewardEvent
---@field kind string -- MatchEventKind or CombatEventKind.
---@field team InputTeam? -- Acting side, when the event has an identified actor.
---@field result CombatContactResult? -- Confirmed combat contact verdict, when applicable.

---@class EnvRewardTransition
---@field team InputTeam -- Perspective the reward is expressed from.
---@field score_before { home: integer, away: integer }
---@field score_after { home: integer, away: integer }
---@field owner_team_before InputTeam?
---@field owner_team_after InputTeam?
---@field events EnvRewardEvent[] -- Confirmed events of the simulated ticks.
---@field terminated boolean -- True only on the tick the match itself ended.

---@class EnvRewardSelection
---@field objectives EnvRewardChannelId[]
---@field shaping EnvRewardChannelId[]

---@class EnvRewardResult
---@field version integer
---@field team InputTeam
---@field objectives table<EnvRewardChannelId, number>
---@field shaping table<EnvRewardChannelId, number>
---@field objective_total number
---@field shaping_total number
---@field total number -- objective_total + shaping_total, for callers that want one scalar.

---@class EnvRewardModule
local env_reward = {}

env_reward.VERSION = 1

-- The banned reward name from #138: predicted human experience is evaluation
-- evidence, never an optimization target.
env_reward.FORBIDDEN_CHANNEL_ID = "fun"

env_reward.DIAGNOSTIC_METRIC_CHANNEL = "experience_proxy_metrics"

local SHOT_EVENT_KINDS = {
    shot = true,
    header = true,
    volley = true,
    bicycle = true,
}

---@type EnvRewardChannel[]
env_reward.CHANNELS = {
    {
        id = "match_outcome",
        role = "objective",
        optimizable = true,
        preregistered = true,
        description = "Sparse +1 win / -1 loss / 0 draw, paid once when the match ends.",
    },
    {
        id = "goal_scored",
        role = "objective",
        optimizable = true,
        preregistered = true,
        description = "+1 per goal scored by the reward team during the step.",
    },
    {
        id = "goal_conceded",
        role = "objective",
        optimizable = true,
        preregistered = true,
        description = "-1 per goal conceded by the reward team during the step.",
    },
    {
        id = "goal_difference_delta",
        role = "objective",
        optimizable = true,
        preregistered = true,
        description = "Change in goal difference from the reward team's perspective.",
    },
    {
        id = "possession_gain",
        role = "shaping",
        optimizable = true,
        preregistered = true,
        description = "+1 when ball ownership transitions to the reward team, -1 when it is lost.",
    },
    {
        id = "shot_attempt",
        role = "shaping",
        optimizable = true,
        preregistered = true,
        description = "+1 per confirmed shot, header, volley, or bicycle strike by the reward team.",
    },
    {
        id = "equipment_contact",
        role = "shaping",
        optimizable = true,
        preregistered = true,
        description = "+1 per unguarded combat contact landed by the reward team.",
    },
    {
        id = "experience_proxy_metrics",
        role = "diagnostic",
        optimizable = false,
        preregistered = true,
        description = "The #128 fun-proxy metric family, returned as evaluation data only.",
    },
}

---@type table<EnvRewardChannelId, EnvRewardChannel>
env_reward.CHANNELS_BY_ID = {}
for _, channel in ipairs(env_reward.CHANNELS) do
    assert(
        channel.id ~= env_reward.FORBIDDEN_CHANNEL_ID,
        "no reward channel may be named " .. env_reward.FORBIDDEN_CHANNEL_ID
    )
    assert(not env_reward.CHANNELS_BY_ID[channel.id], "duplicate reward channel id")
    assert(
        channel.role ~= "diagnostic" or not channel.optimizable,
        "diagnostic reward channels must not be optimizable"
    )
    env_reward.CHANNELS_BY_ID[channel.id] = channel
end

---@type EnvRewardChannelId[]
env_reward.DEFAULT_OBJECTIVES = { "match_outcome" }

---@param code EnvRewardErrorCode
---@param message string
---@return nil, string, EnvRewardErrorCode
local function reject(code, message)
    return nil, message, code
end

---@param value any
---@return boolean
local function is_side(value)
    return value == "home" or value == "away"
end

-- Validate an authored channel selection for one role. Selections come from a
-- run config (external input), so failures are recoverable returns.
---@param ids any
---@param role EnvRewardRole
---@return EnvRewardChannelId[]?, string?, EnvRewardErrorCode?
function env_reward.validate_selection(ids, role)
    assert(role == "objective" or role == "shaping", "unknown reward role: " .. tostring(role))
    if type(ids) ~= "table" then
        return reject("malformed", role .. " channel selection must be an array")
    end
    local count = #ids
    for key in pairs(ids) do
        if type(key) ~= "number" or key ~= math.floor(key) or key < 1 or key > count then
            return reject("malformed", role .. " channel selection is not a canonical array")
        end
    end
    local seen = {}
    local copied = {}
    for index = 1, count do
        local id = ids[index]
        if type(id) ~= "string" then
            return reject("malformed", role .. " channel id must be a string")
        end
        local channel = env_reward.CHANNELS_BY_ID[id]
        if not channel then
            return reject("unknown_channel", "unknown reward channel: " .. id)
        end
        if channel.role ~= role then
            return reject(
                "wrong_role",
                "reward channel " .. id .. " is a " .. channel.role .. " channel, not " .. role
            )
        end
        if seen[id] then
            return reject("duplicate_channel", "reward channel " .. id .. " is selected twice")
        end
        seen[id] = true
        copied[index] = id
    end
    return copied
end

---@param selection any
---@return EnvRewardSelection?, string?, EnvRewardErrorCode?
function env_reward.validate(selection)
    if type(selection) ~= "table" then
        return reject("malformed", "reward selection must be a table")
    end
    for key in pairs(selection) do
        if key ~= "objectives" and key ~= "shaping" then
            return reject("malformed", "reward selection has unknown field " .. tostring(key))
        end
    end
    local objectives, objectives_err, objectives_code =
        env_reward.validate_selection(selection.objectives or {}, "objective")
    if not objectives then
        return nil, objectives_err, objectives_code
    end
    local shaping, shaping_err, shaping_code =
        env_reward.validate_selection(selection.shaping or {}, "shaping")
    if not shaping then
        return nil, shaping_err, shaping_code
    end
    return { objectives = objectives, shaping = shaping }
end

---@param transition EnvRewardTransition
---@return integer own_goals
---@return integer opponent_goals
local function goal_deltas(transition)
    local before, after = transition.score_before, transition.score_after
    local own_before = transition.team == "home" and before.home or before.away
    local own_after = transition.team == "home" and after.home or after.away
    local opponent_before = transition.team == "home" and before.away or before.home
    local opponent_after = transition.team == "home" and after.away or after.home
    return own_after - own_before, opponent_after - opponent_before
end

---@param transition EnvRewardTransition
---@return number
local function match_outcome(transition)
    if not transition.terminated then
        return 0
    end
    local after = transition.score_after
    local own = transition.team == "home" and after.home or after.away
    local opponent = transition.team == "home" and after.away or after.home
    if own > opponent then
        return 1
    end
    if own < opponent then
        return -1
    end
    return 0
end

---@param transition EnvRewardTransition
---@return number
local function possession_gain(transition)
    local before = transition.owner_team_before == transition.team
    local after = transition.owner_team_after == transition.team
    if after and not before then
        return 1
    end
    if before and not after then
        return -1
    end
    return 0
end

---@param transition EnvRewardTransition
---@return number
local function shot_attempt(transition)
    local total = 0
    for _, event in ipairs(transition.events) do
        if SHOT_EVENT_KINDS[event.kind] and event.team == transition.team then
            total = total + 1
        end
    end
    return total
end

---@param transition EnvRewardTransition
---@return number
local function equipment_contact(transition)
    local total = 0
    for _, event in ipairs(transition.events) do
        if event.kind == "contact" and event.team == transition.team and event.result == "hit" then
            total = total + 1
        end
    end
    return total
end

---@type table<EnvRewardChannelId, fun(transition: EnvRewardTransition): number>
local EVALUATORS = {
    match_outcome = match_outcome,
    goal_scored = function(transition)
        local own = goal_deltas(transition)
        return own
    end,
    goal_conceded = function(transition)
        local _, opponent = goal_deltas(transition)
        return -opponent
    end,
    goal_difference_delta = function(transition)
        local own, opponent = goal_deltas(transition)
        return own - opponent
    end,
    possession_gain = possession_gain,
    shot_attempt = shot_attempt,
    equipment_contact = equipment_contact,
}

---@param transition EnvRewardTransition
local function assert_transition(transition)
    assert(type(transition) == "table", "reward transition is required")
    assert(is_side(transition.team), "reward transition needs a home/away perspective")
    assert(type(transition.score_before) == "table", "reward transition needs score_before")
    assert(type(transition.score_after) == "table", "reward transition needs score_after")
    assert(type(transition.events) == "table", "reward transition needs an event array")
    assert(type(transition.terminated) == "boolean", "reward transition needs a terminated flag")
    assert(
        transition.owner_team_before == nil or is_side(transition.owner_team_before),
        "reward transition owner_team_before must be a side or nil"
    )
    assert(
        transition.owner_team_after == nil or is_side(transition.owner_team_after),
        "reward transition owner_team_after must be a side or nil"
    )
end

-- Score one already-simulated transition. Objectives and shaping are summed
-- separately so an ablation report can drop the shaping column without
-- re-running the episode.
---@param transition EnvRewardTransition
---@param selection EnvRewardSelection
---@return EnvRewardResult
function env_reward.evaluate(transition, selection)
    assert_transition(transition)
    assert(type(selection) == "table", "reward selection is required")
    local objectives = {}
    local shaping = {}
    local objective_total = 0
    local shaping_total = 0
    for _, id in ipairs(selection.objectives) do
        local channel = assert(env_reward.CHANNELS_BY_ID[id], "unknown objective channel: " .. id)
        assert(channel.role == "objective", "channel " .. id .. " is not an objective")
        local value =
            assert(EVALUATORS[id], "objective channel has no evaluator: " .. id)(transition)
        objectives[id] = value
        objective_total = objective_total + value
    end
    for _, id in ipairs(selection.shaping) do
        local channel = assert(env_reward.CHANNELS_BY_ID[id], "unknown shaping channel: " .. id)
        assert(channel.role == "shaping", "channel " .. id .. " is not shaping")
        local value = assert(EVALUATORS[id], "shaping channel has no evaluator: " .. id)(transition)
        shaping[id] = value
        shaping_total = shaping_total + value
    end
    return {
        version = env_reward.VERSION,
        team = transition.team,
        objectives = objectives,
        shaping = shaping,
        objective_total = objective_total,
        shaping_total = shaping_total,
        total = objective_total + shaping_total,
    }
end

return env_reward

File: rust/crates/gc-sim/src/env_reward.rs
//! Named learning-environment reward channels.
//!
//! There is no implicit scalar reward. Every channel is registered with an
//! explicit role: `objective` channels are competition targets, `shaping`
//! channels must be preregistered and are always reported separately so an
//! ablation is a subtraction rather than a rerun, and `diagnostic` channels
//! are evaluation data that may never be optimized. The #128 fun-proxy
//! metric family is a diagnostic channel: selecting it as an objective or
//! shaping term is a validation error, and no channel may be named `fun`.
//!
//! This module has no `sim` dependencies of its own: the fixture side
//! ([`EnvSide`]) and confirmed combat verdict ([`CombatContactResult`])
//! types it references are small closed sets duplicated locally rather than
//! imported from `input_frame`/`combat`, keeping this module's dependency
//! surface minimal.

use indexmap::IndexMap;

/// The env reward module version.
pub const VERSION: i64 = 1;

/// The banned reward name from #138: predicted human experience is
/// evaluation evidence, never an optimization target. No [`EnvRewardChannelId`]
/// may spell this name; enforced by construction since the id is a closed
/// enum, and checked by name in the registry self-test.
pub const FORBIDDEN_CHANNEL_NAME: &str = "fun";

/// The #128 fun-proxy metric family: diagnostic, evaluation-only.
pub const DIAGNOSTIC_METRIC_CHANNEL: EnvRewardChannelId =
    EnvRewardChannelId::ExperienceProxyMetrics;

/// The default objective selection: the sparse match outcome only.
pub const DEFAULT_OBJECTIVES: [EnvRewardChannelId; 1] = [EnvRewardChannelId::MatchOutcome];

const SHOT_EVENT_KINDS: [&str; 4] = ["shot", "header", "volley", "bicycle"];

/// A fixture side. Declared locally rather than reusing
/// [`crate::input_frame::Team`], keeping this module's dependency surface
/// minimal — it checks side membership without needing `input_frame`.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum EnvSide {
    /// Home side.
    Home,
    /// Away side.
    Away,
}

/// A confirmed combat contact verdict. Declared locally rather than reusing
/// [`crate::combat_snapshot::CombatContactResult`], keeping this module's
/// dependency surface minimal.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CombatContactResult {
    /// Landed cleanly.
    Hit,
    /// Landed past a partial guard.
    Extended,
    /// Blocked by a guard.
    Guarded,
    /// No-effect due to immunity frames.
    Immune,
    /// Superseded by a later, higher-priority action.
    Superseded,
}

/// A reward channel's declared role in an episode.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum EnvRewardRole {
    /// A competition target.
    Objective,
    /// A preregistered, separately-reported shaping term.
    Shaping,
    /// Evaluation data that may never be optimized.
    Diagnostic,
}

fn role_name(role: EnvRewardRole) -> &'static str {
    match role {
        EnvRewardRole::Objective => "objective",
        EnvRewardRole::Shaping => "shaping",
        EnvRewardRole::Diagnostic => "diagnostic",
    }
}

/// The closed set of registered reward channel ids.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum EnvRewardChannelId {
    /// Sparse +1 win / -1 loss / 0 draw, paid once when the match ends.
    MatchOutcome,
    /// +1 per goal scored by the reward team during the step.
    GoalScored,
    /// -1 per goal conceded by the reward team during the step.
    GoalConceded,
    /// Change in goal difference from the reward team's perspective.
    GoalDifferenceDelta,
    /// +1 when ball ownership transitions to the reward team, -1 when lost.
    PossessionGain,
    /// +1 per confirmed shot, header, volley, or bicycle strike by the
    /// reward team.
    ShotAttempt,
    /// +1 per unguarded combat contact landed by the reward team.
    EquipmentContact,
    /// The #128 fun-proxy metric family, returned as evaluation data only.
    ExperienceProxyMetrics,
}

/// The channel id's canonical string form, as it appears on the wire (a
/// config digest, a serialized selection).
#[must_use]
pub fn channel_id_name(id: EnvRewardChannelId) -> &'static str {
    match id {
        EnvRewardChannelId::MatchOutcome => "match_outcome",
        EnvRewardChannelId::GoalScored => "goal_scored",
        EnvRewardChannelId::GoalConceded => "goal_conceded",
        EnvRewardChannelId::GoalDifferenceDelta => "goal_difference_delta",
        EnvRewardChannelId::PossessionGain => "possession_gain",
        EnvRewardChannelId::ShotAttempt => "shot_attempt",
        EnvRewardChannelId::EquipmentContact => "equipment_contact",
        EnvRewardChannelId::ExperienceProxyMetrics => "experience_proxy_metrics",
    }
}

/// The reverse of [`channel_id_name`].
#[must_use]
pub fn channel_id_from_name(name: &str) -> Option<EnvRewardChannelId> {
    ALL_CHANNEL_IDS
        .into_iter()
        .find(|id| channel_id_name(*id) == name)
}

const ALL_CHANNEL_IDS: [EnvRewardChannelId; 8] = [
    EnvRewardChannelId::MatchOutcome,
    EnvRewardChannelId::GoalScored,
    EnvRewardChannelId::GoalConceded,
    EnvRewardChannelId::GoalDifferenceDelta,
    EnvRewardChannelId::PossessionGain,
    EnvRewardChannelId::ShotAttempt,
    EnvRewardChannelId::EquipmentContact,
    EnvRewardChannelId::ExperienceProxyMetrics,
];

/// One registered reward channel's declared identity.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct EnvRewardChannel {
    /// The channel's id.
    pub id: EnvRewardChannelId,
    /// The channel's declared role.
    pub role: EnvRewardRole,
    /// False for diagnostics: never a training target.
    pub optimizable: bool,
    /// Shaping terms are declared here, not invented per run.
    pub preregistered: bool,
    /// Human-readable description.
    pub description: &'static str,
}

/// Every registered reward channel, in canonical order.
pub const CHANNELS: [EnvRewardChannel; 8] = [
    EnvRewardChannel {
        id: EnvRewardChannelId::MatchOutcome,
        role: EnvRewardRole::Objective,
        optimizable: true,
        preregistered: true,
        description: "Sparse +1 win / -1 loss / 0 draw, paid once when the match ends.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::GoalScored,
        role: EnvRewardRole::Objective,
        optimizable: true,
        preregistered: true,
        description: "+1 per goal scored by the reward team during the step.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::GoalConceded,
        role: EnvRewardRole::Objective,
        optimizable: true,
        preregistered: true,
        description: "-1 per goal conceded by the reward team during the step.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::GoalDifferenceDelta,
        role: EnvRewardRole::Objective,
        optimizable: true,
        preregistered: true,
        description: "Change in goal difference from the reward team's perspective.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::PossessionGain,
        role: EnvRewardRole::Shaping,
        optimizable: true,
        preregistered: true,
        description: "+1 when ball ownership transitions to the reward team, -1 when it is lost.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::ShotAttempt,
        role: EnvRewardRole::Shaping,
        optimizable: true,
        preregistered: true,
        description: "+1 per confirmed shot, header, volley, or bicycle strike by the reward team.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::EquipmentContact,
        role: EnvRewardRole::Shaping,
        optimizable: true,
        preregistered: true,
        description: "+1 per unguarded combat contact landed by the reward team.",
    },
    EnvRewardChannel {
        id: EnvRewardChannelId::ExperienceProxyMetrics,
        role: EnvRewardRole::Diagnostic,
        optimizable: false,
        preregistered: true,
        description: "The #128 fun-proxy metric family, returned as evaluation data only.",
    },
];

/// Look up a registered channel's full declaration by id.
#[must_use]
pub fn channel(id: EnvRewardChannelId) -> EnvRewardChannel {
    CHANNELS
        .into_iter()
        .find(|channel| channel.id == id)
        .expect("every EnvRewardChannelId is registered in CHANNELS")
}

/// Failure reasons an env-reward operation can report.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum EnvRewardErrorCode {
    /// The data violates a structural or type invariant.
    Malformed,
    /// A channel id names something outside the registry.
    UnknownChannel,
    /// A channel id is registered, but not for the requested role.
    WrongRole,
    /// A channel id was selected twice for the same role.
    DuplicateChannel,
}

/// An expected, recoverable env-reward failure (ARCHITECTURE.md §3 rule 5): a channel
/// selection comes from a run config (external input), so a bad selection is
/// a recoverable rejection with a machine-readable reason, never a panic.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EnvRewardError {
    /// Machine-readable failure reason.
    pub code: EnvRewardErrorCode,
    /// Human-readable detail.
    pub message: String,
}

impl EnvRewardError {
    fn new(code: EnvRewardErrorCode, message: impl Into<String>) -> Self {
        EnvRewardError {
            code,
            message: message.into(),
        }
    }
}

impl std::fmt::Display for EnvRewardError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for EnvRewardError {}

/// Result alias for fallible env-reward operations.
pub type Result<T> = std::result::Result<T, EnvRewardError>;

fn failure<T>(code: EnvRewardErrorCode, message: impl Into<String>) -> Result<T> {
    Err(EnvRewardError::new(code, message))
}

// ---------------------------------------------------------------------------
// Raw, not-yet-validated external input.
//
// A channel selection arrives from a run config, so `validate_selection` and
// `validate` accept a genuinely untyped shape rather than assuming it is
// already well-formed.
// ---------------------------------------------------------------------------

/// One raw, not-yet-validated channel id entry.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum RawChannelIdEntry {
    /// A string id — the only legal shape.
    Str(String),
    /// Anything else.
    Other,
}

/// A raw, not-yet-validated channel selection list for one role.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum RawChannelIds {
    /// An ordered list of raw ids.
    List(Vec<RawChannelIdEntry>),
    /// Anything that is not a list.
    Other,
}

/// Validate an authored channel selection for one role. Selections come from
/// a run config (external input), so failures are recoverable returns.
pub fn validate_selection(
    ids: &RawChannelIds,
    role: EnvRewardRole,
) -> Result<Vec<EnvRewardChannelId>> {
    let entries = match ids {
        RawChannelIds::List(entries) => entries,
        RawChannelIds::Other => {
            return failure(
                EnvRewardErrorCode::Malformed,
                format!("{} channel selection must be an array", role_name(role)),
            );
        }
    };
    let mut seen: Vec<EnvRewardChannelId> = Vec::new();
    let mut copied: Vec<EnvRewardChannelId> = Vec::new();
    for entry in entries {
        let RawChannelIdEntry::Str(id) = entry else {
            return failure(
                EnvRewardErrorCode::Malformed,
                format!("{} channel id must be a string", role_name(role)),
            );
        };
        let Some(channel_id) = channel_id_from_name(id) else {
            return failure(
                EnvRewardErrorCode::UnknownChannel,
                format!("unknown reward channel: {id}"),
            );
        };
        let declared = channel(channel_id);
        if declared.role != role {
            return failure(
                EnvRewardErrorCode::WrongRole,
                format!(
                    "reward channel {} is a {} channel, not {}",
                    id,
                    role_name(declared.role),
                    role_name(role)
                ),
            );
        }
        if seen.contains(&channel_id) {
            return failure(
                EnvRewardErrorCode::DuplicateChannel,
                format!("reward channel {id} is selected twice"),
            );
        }
        seen.push(channel_id);
        copied.push(channel_id);
    }
    Ok(copied)
}

/// A raw, not-yet-validated top-level reward-channel selection, exactly as
/// received from run configuration.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum RawSelection {
    /// A table naming only `objectives`/`shaping`.
    Table(RawSelectionTable),
    /// Anything that is not a table.
    Other,
}

/// The two legal fields of a [`RawSelection::Table`].
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct RawSelectionTable {
    /// The raw `objectives` field, when present.
    pub objectives: Option<RawChannelIds>,
    /// The raw `shaping` field, when present.
    pub shaping: Option<RawChannelIds>,
    /// True when the caller's table carried a field other than the two
    /// above.
    pub has_unknown_field: bool,
}

/// A validated, complete reward-channel selection.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct EnvRewardSelection {
    /// Selected objective channel ids, in caller order.
    pub objectives: Vec<EnvRewardChannelId>,
    /// Selected shaping channel ids, in caller order.
    pub shaping: Vec<EnvRewardChannelId>,
}

/// Validate a caller's whole reward-channel selection.
pub fn validate(selection: &RawSelection) -> Result<EnvRewardSelection> {
    let table = match selection {
        RawSelection::Table(table) => table,
        RawSelection::Other => {
            return failure(
                EnvRewardErrorCode::Malformed,
                "reward selection must be a table",
            );
        }
    };
    if table.has_unknown_field {
        return failure(
            EnvRewardErrorCode::Malformed,
            "reward selection has unknown field",
        );
    }
    let empty = RawChannelIds::List(Vec::new());
    let objectives = validate_selection(
        table.objectives.as_ref().unwrap_or(&empty),
        EnvRewardRole::Objective,
    )?;
    let shaping = validate_selection(
        table.shaping.as_ref().unwrap_or(&empty),
        EnvRewardRole::Shaping,
    )?;
    Ok(EnvRewardSelection {
        objectives,
        shaping,
    })
}

// ---------------------------------------------------------------------------
// Evaluation.
// ---------------------------------------------------------------------------

/// Home/away goal counts at one instant.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub struct EnvRewardScore {
    /// Home goals.
    pub home: i64,
    /// Away goals.
    pub away: i64,
}

/// One confirmed match/combat event, as far as reward evaluation needs it.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EnvRewardEvent {
    /// `MatchEventKind` or `CombatEventKind`, as a bare string: this field
    /// can hold either kind depending on the event's source, and a plain
    /// `String` avoids inventing a third enum that is just a union of the
    /// other two.
    pub kind: String,
    /// Acting side, when the event has an identified actor.
    pub team: Option<EnvSide>,
    /// Confirmed combat contact verdict, when applicable.
    pub result: Option<CombatContactResult>,
}

/// One already-simulated transition, expressed from one team's perspective.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EnvRewardTransition {
    /// Perspective the reward is expressed from.
    pub team: EnvSide,
    /// Score immediately before the simulated ticks.
    pub score_before: EnvRewardScore,
    /// Score immediately after the simulated ticks.
    pub score_after: EnvRewardScore,
    /// Ball owner before the simulated ticks, if any.
    pub owner_team_before: Option<EnvSide>,
    /// Ball owner after the simulated ticks, if any.
    pub owner_team_after: Option<EnvSide>,
    /// Confirmed events of the simulated ticks.
    pub events: Vec<EnvRewardEvent>,
    /// True only on the tick the match itself ended.
    pub terminated: bool,
}

/// One evaluated reward: every selected channel, plus the objective/shaping
/// subtotals and their sum.
#[derive(Clone, Debug, PartialEq)]
pub struct EnvRewardResult {
    /// Exactly [`VERSION`].
    pub version: i64,
    /// Perspective the reward is expressed from.
    pub team: EnvSide,
    /// Per-channel objective values, only for selected channels.
    pub objectives: IndexMap<EnvRewardChannelId, f64>,
    /// Per-channel shaping values, only for selected channels.
    pub shaping: IndexMap<EnvRewardChannelId, f64>,
    /// Sum of `objectives`.
    pub objective_total: f64,
    /// Sum of `shaping`.
    pub shaping_total: f64,
    /// `objective_total + shaping_total`, for callers that want one scalar.
    pub total: f64,
}

fn perspective(team: EnvSide, score: &EnvRewardScore) -> (i64, i64) {
    match team {
        EnvSide::Home => (score.home, score.away),
        EnvSide::Away => (score.away, score.home),
    }
}

fn goal_deltas(transition: &EnvRewardTransition) -> (i64, i64) {
    let (own_before, opponent_before) = perspective(transition.team, &transition.score_before);
    let (own_after, opponent_after) = perspective(transition.team, &transition.score_after);
    (own_after - own_before, opponent_after - opponent_before)
}

fn match_outcome(transition: &EnvRewardTransition) -> f64 {
    if !transition.terminated {
        return 0.0;
    }
    let (own, opponent) = perspective(transition.team, &transition.score_after);
    if own > opponent {
        1.0
    } else if own < opponent {
        -1.0
    } else {
        0.0
    }
}

fn possession_gain(transition: &EnvRewardTransition) -> f64 {
    let before = transition.owner_team_before == Some(transition.team);
    let after = transition.owner_team_after == Some(transition.team);
    if after && !before {
        1.0
    } else if before && !after {
        -1.0
    } else {
        0.0
    }
}

fn shot_attempt(transition: &EnvRewardTransition) -> f64 {
    let mut total = 0.0;
    for event in &transition.events {
        if SHOT_EVENT_KINDS.contains(&event.kind.as_str()) && event.team == Some(transition.team) {
            total += 1.0;
        }
    }
    total
}

fn equipment_contact(transition: &EnvRewardTransition) -> f64 {
    let mut total = 0.0;
    for event in &transition.events {
        if event.kind == "contact"
            && event.team == Some(transition.team)
            && event.result == Some(CombatContactResult::Hit)
        {
            total += 1.0;
        }
    }
    total
}

fn evaluate_channel(id: EnvRewardChannelId, transition: &EnvRewardTransition) -> f64 {
    match id {
        EnvRewardChannelId::MatchOutcome => match_outcome(transition),
        EnvRewardChannelId::GoalScored => {
            let (own, _) = goal_deltas(transition);
            own as f64
        }
        EnvRewardChannelId::GoalConceded => {
            let (_, opponent) = goal_deltas(transition);
            -(opponent as f64)
        }
        EnvRewardChannelId::GoalDifferenceDelta => {
            let (own, opponent) = goal_deltas(transition);
            (own - opponent) as f64
        }
        EnvRewardChannelId::PossessionGain => possession_gain(transition),
        EnvRewardChannelId::ShotAttempt => shot_attempt(transition),
        EnvRewardChannelId::EquipmentContact => equipment_contact(transition),
        EnvRewardChannelId::ExperienceProxyMetrics => {
            unreachable!(
                "experience_proxy_metrics is diagnostic; validate_selection never admits it \
                 into an objective/shaping selection"
            )
        }
    }
}

/// Score one already-simulated transition. Objectives and shaping are summed
/// separately so an ablation report can drop the shaping column without
/// re-running the episode.
///
/// `transition` and `selection` are already-typed values, not raw external
/// input; a `selection` naming an unregistered or wrong-role channel is a
/// programmer error caught by `assert` (AGENTS.md §7), not a recoverable
/// rejection — [`validate_selection`] is what external callers must run
/// first.
#[must_use]
pub fn evaluate(
    transition: &EnvRewardTransition,
    selection: &EnvRewardSelection,
) -> EnvRewardResult {
    let mut objectives = IndexMap::new();
    let mut shaping = IndexMap::new();
    let mut objective_total = 0.0;
    let mut shaping_total = 0.0;
    for &id in &selection.objectives {
        let declared = channel(id);
        assert_eq!(
            declared.role,
            EnvRewardRole::Objective,
            "channel {id:?} is not an objective"
        );
        let value = evaluate_channel(id, transition);
        objectives.insert(id, value);
        objective_total += value;
    }
    for &id in &selection.shaping {
        let declared = channel(id);
        assert_eq!(
            declared.role,
            EnvRewardRole::Shaping,
            "channel {id:?} is not shaping"
        );
        let value = evaluate_channel(id, transition);
        shaping.insert(id, value);
        shaping_total += value;
    }
    EnvRewardResult {
        version: VERSION,
        team: transition.team,
        objectives,
        shaping,
        objective_total,
        shaping_total,
        total: objective_total + shaping_total,
    }
}
